utils/frame/benchmarking-cli/src/writer.rs

```diff
@@ -17,314 +17,401 @@
-		write!(file, "{}fn {}(", indent, benchmark_string)?;
-
-		// params
-		let components = &batch.results[0].components;
-		for component in components {
-			write!(file, "{:?}: u32, ", component.0)?;
+// Analyze and return the relevant results for a given benchmark.
+fn get_benchmark_data(batch: &BenchmarkBatch) -> BenchmarkData {
+	// Analyze benchmarks to get the linear regression.
+	let extrinsic_time = Analysis::min_squares_iqr(&batch.results, BenchmarkSelector::ExtrinsicTime).unwrap();
+	let reads = Analysis::min_squares_iqr(&batch.results, BenchmarkSelector::Reads).unwrap();
+	let writes = Analysis::min_squares_iqr(&batch.results, BenchmarkSelector::Writes).unwrap();
+
+	// Analysis data may include components that are not used, this filters out anything whose value is zero.
+	let mut used_components = Vec::new();
+	let mut used_extrinsic_time = Vec::new();
+	let mut used_reads = Vec::new();
+	let mut used_writes = Vec::new();
+
+	extrinsic_time.slopes.into_iter().zip(extrinsic_time.names.iter()).for_each(|(slope, name)| {
+		if !slope.is_zero() {
+			if !used_components.contains(&name) { used_components.push(name); }
+			used_extrinsic_time.push(ComponentSlope {
+				name: name.clone(),
+				slope: slope.saturating_mul(1000),
+			});
 		}
-		// return value
-		write!(file, ") -> Weight;\n")?;
+	});
+	reads.slopes.into_iter().zip(reads.names.iter()).for_each(|(slope, name)| {
+		if !slope.is_zero() {
+			if !used_components.contains(&name) { used_components.push(name); }
+			used_reads.push(ComponentSlope { name: name.clone(), slope });
+		}
+	});
+	writes.slopes.into_iter().zip(writes.names.iter()).for_each(|(slope, name)| {
+		if !slope.is_zero() {
+			if !used_components.contains(&name) { used_components.push(name); }
+			used_writes.push(ComponentSlope { name: name.clone(), slope });
+		}
+	});
+
+	// This puts a marker on any component which is entirely unused in the weight formula.
+	let components = batch.results[0].components
+		.iter()
+		.map(|(name, _)| -> Component {
+			let name_string = name.to_string();
+			let is_used = used_components.contains(&&name_string);
+			Component { name: name_string, is_used }
+		})
+		.collect::<Vec<_>>();
+
+	BenchmarkData {
+		name: String::from_utf8(batch.benchmark.clone()).unwrap(),
+		components,
+		base_weight: extrinsic_time.base.saturating_mul(1000),
+		base_reads: reads.base,
+		base_writes: writes.base,
+		component_weight: used_extrinsic_time,
+		component_reads: used_reads,
+		component_writes: used_writes,
 	}
-
-	// final close trait
-	write!(file, "}}\n")?;
-
-	Ok(())
 }
 
+// Create weight file from benchmark data and Handlebars template.
 pub fn write_results(
 	batches: &[BenchmarkBatch],
 	path: &PathBuf,
 	cmd: &BenchmarkCmd,
 ) -> Result<(), std::io::Error> {
+	// Use custom template if provided.
+	let template: String = match &cmd.template {
+		Some(template_file) => {
+			fs::read_to_string(template_file)?
+		},
+		None => {
+			TEMPLATE.to_string()
+		},
+	};
```

Comment:
I think it better to use the default value feature of structopt than this custom logic. By doing that the default template is automatically discoverable on the CLI.